File: bwell-swift-ios/bwell-swift-ios/HealthSync/HealthSyncPlaygroundViewModel.swift
//
//  HealthSyncPlaygroundViewModel.swift
//  bwell-swift-ios
//
//  Adapted from bwell-sdk-swift/Examples/HealthSyncSampleApp/Sources/HealthSyncViewModel.swift.
//  Unlike the internal sample, this demo does not own its own BWellClient or
//  login flow - it attaches to the app's existing SDKManager session (see
//  HealthSyncPlaygroundRootView.swift). Whether the visitor is the current
//  logged-in user is the app's own concern, decided before this screen is
//  ever reachable, not something re-checked here.
//
//  Vendor-neutrality: `AggregatorHealthSource` (the on-device health source
//  implementation) lives in a package that pulls a third-party vendor SDK
//  through b.well's private Artifactory registry — this public repo cannot
//  resolve it without credentials, so it is never added as a dependency here.
//  The `#if canImport(...)` guard below means `BWellHealthSyncType.configure`
//  is simply never called in this build: every on-device card's
//  `client.healthSync` access throws the SDK's own vendor-neutral
//  `HealthSyncError.notConfigured` (see `runPlaygroundEndpoint`), which this
//  file maps to a plain-language fallback. A developer with real credentials
//  can add the adapter package locally via Xcode; this exact same source
//  then activates automatically, no further edits needed.
//

import BWellHealthSync
#if canImport(BWellHealthSyncAdapterAggregator)
import BWellHealthSyncAdapterAggregator
#endif
import BWellSDK
import Foundation

/// A dynamically-fetched dropdown option identified by a code, with an
/// optional human-readable meaning (e.g. code "vital-signs", display "Vital
/// Signs"). Shared shape for both the getDeviceMetrics code picker and the
/// getBodySystemScore body-system picker - both are "code - meaning" pairs
/// sourced live from another endpoint's response, not hardcoded.
struct PlaygroundCodeOption: Identifiable, Hashable {
    let code: String
    let display: String?
    var id: String { code }
    var label: String { display.map { "\(code) - \($0)" } ?? code }
}

@MainActor
final class HealthSyncPlaygroundViewModel: ObservableObject {
    @Published var playgroundResults: [HealthSyncPlaygroundEndpoint: PlaygroundCardState] = [:]
    @Published var deviceProviderOptions: [BWell.DeviceProvider] = []
    @Published var selectedDeviceProvider: BWell.DeviceProvider?
    @Published var connectionIdInput = playgroundConnectionId
    @Published var deviceMetricsCodeOptions: [PlaygroundCodeOption] = []
    @Published var selectedDeviceMetricsCode: String?
    @Published var bodySystemOptions: [PlaygroundCodeOption] = []
    @Published var selectedBodySystemId: String?

    private var client: BWellClient?

    // Not SDK surface (sourceIdToConnectionSlug is internal) — hardcoded here
    // the same way a real integrator would, since the demo only ever
    // demonstrates this one source (AggregatorHealthSource.sourceId = "apple-health").
    static let playgroundConnectionId = "apple_health"

    // How far back to read on-device records for the demo `sync()` card — a
    // demo-app choice (not an SDK default), picked to reliably surface
    // something to show on real devices without a huge upload.
    private static let syncWindowDays = 30
    private static let secondsPerDay: TimeInterval = 86400

    /// Attaches the app's already-authenticated client - called once from
    /// HealthSyncPlaygroundRootView, which owns the actual session. Also
    /// where the on-device health source gets configured, when the adapter
    /// package is present (see file header).
    func attach(client: BWellClient) {
        guard self.client == nil else { return }
        self.client = client

        #if canImport(BWellHealthSyncAdapterAggregator)
        if !BWellHealthSyncType.isConfigured {
            try? BWellHealthSyncType.configure(AggregatorHealthSource())
        }
        #endif

        loadPlaygroundDropdownOptions()
    }

    // MARK: - Playground

    /// Silently (re-)populates the provider dropdown from its natural parent
    /// endpoint. Failures are silent - this is a dev-tooling convenience,
    /// not the endpoint the user is actually testing.
    private func loadPlaygroundDropdownOptions() {
        Task {
            do {
                if let result = try await client?.device.getDeviceProviders() {
                    deviceProviderOptions = result.providers
                    if selectedDeviceProvider == nil {
                        selectedDeviceProvider = result.providers.first
                    }
                }
            } catch {
                print("getDeviceProviders() dropdown prefetch failed: \(error)")
            }
        }
    }

    /// Re-populates both dropdowns sourced from another endpoint's live
    /// response - called right after a successful `sync` (see
    /// runPlaygroundEndpoint) so newly-synced data shows up on this same
    /// screen without waiting for the picker's own `.task` to remount.
    /// Mirrors Kotlin's HealthSyncPlaygroundViewModel.refreshDeviceDataOptions().
    func refreshDeviceDataOptions() {
        loadDeviceMetricsCodeOptions()
        loadBodySystemOptions()
    }

    /// (Re-)populates the getDeviceMetrics code dropdown from
    /// getDeviceMetricsGroups() - called each time that card appears, since
    /// which codes exist depends on what's been synced (Mobile Sync group
    /// above), which can change after this screen's initial load.
    func loadDeviceMetricsCodeOptions() {
        Task {
            do {
                guard let result = try await client?.health.getDeviceMetricsGroups(nil) else { return }
                var seen = Set<String>()
                let options = (result.resources ?? []).compactMap { group -> PlaygroundCodeOption? in
                    guard let code = group.coding?.code, seen.insert(code).inserted else { return nil }
                    return PlaygroundCodeOption(code: code, display: group.coding?.display)
                }.sorted { $0.code < $1.code }
                deviceMetricsCodeOptions = options
                if let selectedDeviceMetricsCode, !options.contains(where: { $0.code == selectedDeviceMetricsCode }) {
                    self.selectedDeviceMetricsCode = nil
                }
            } catch {
                print("getDeviceMetricsGroups() dropdown prefetch failed: \(error)")
            }
        }
    }

    /// (Re-)populates the getBodySystemScore body-system dropdown from
    /// getHealthScore()'s contributing body systems - called each time that
    /// card appears, since which body systems have a score depends on how
    /// much has been synced, which can change after this screen's initial load.
    func loadBodySystemOptions() {
        Task {
            do {
                guard let result = try await client?.health.getHealthScore() else { return }
                var seen = Set<String>()
                let options = (result.resource?.bodySystems ?? []).compactMap { summary -> PlaygroundCodeOption? in
                    guard let id = summary.bodySystemId, seen.insert(id).inserted else { return nil }
                    return PlaygroundCodeOption(code: id, display: summary.title)
                }.sorted { $0.code < $1.code }
                bodySystemOptions = options
                if let selectedBodySystemId, !options.contains(where: { $0.code == selectedBodySystemId }) {
                    self.selectedBodySystemId = nil
                }
                // Unlike the device-metrics code picker, there's no "All"
                // option here - getBodySystemScore always needs exactly one
                // id, so default to the first one once options load.
                if selectedBodySystemId == nil {
                    selectedBodySystemId = options.first?.code
                }
            } catch {
                print("getHealthScore() dropdown prefetch failed: \(error)")
            }
        }
    }

    /// Runs one playground endpoint independently — each card has its own
    /// loading/result state.
    func runPlaygroundEndpoint(_ endpoint: HealthSyncPlaygroundEndpoint) {
        if let reason = endpoint.blockedReason {
            playgroundResults[endpoint] = .error(reason)
            return
        }
        playgroundResults[endpoint] = .loading
        Task {
            guard let client else {
                playgroundResults[endpoint] = .error("SDK not initialized")
                return
            }
            do {
                let description = try await run(endpoint, client: client)
                playgroundResults[endpoint] = .success(description)
                if endpoint == .sync {
                    refreshDeviceDataOptions()
                }
            } catch PlaygroundInputError.missingInput(let message) {
                playgroundResults[endpoint] = .error(message)
            } catch HealthSyncError.notConfigured {
                // Shouldn't be reachable - the 5 gated endpoints' Run
                // buttons are pre-emptively disabled via isBlocked/
                // blockedReason above. Kept as a defensive fallback, same
                // copy as that locked-state message.
                playgroundResults[endpoint] = .error(HealthSyncPlaygroundEndpoint.notConfiguredMessage)
            } catch {
                playgroundResults[endpoint] = .error("\(error)")
            }
        }
    }

    private enum PlaygroundInputError: Error {
        case missingInput(String)
    }

    private func run(_ endpoint: HealthSyncPlaygroundEndpoint, client: BWellClient) async throws -> String {
        switch endpoint {
        case .connect:
            try await reconcileHealthSyncSessionIfNeeded(client: client, connectionId: Self.playgroundConnectionId)
            try await client.healthSync.connect()
            return "connect() completed successfully."

        case .disconnect:
            try await client.healthSync.disconnect()
            return "disconnect() completed successfully."

        case .getCurrentUser:
            guard let user = try await client.healthSync.currentUser() else {
                return "No active session."
            }
            return "userId: \(user.userId)\norganizationId: \(user.organizationId)"

        case .requestPermissions:
            try await client.healthSync.requestPermissions(Set(HealthDataType.allCases))
            return "requestPermissions() completed successfully for \(HealthDataType.allCases.count) types."

        case .sync:
            let now = Date()
            let windowStart = now.addingTimeInterval(-Double(Self.syncWindowDays) * Self.secondsPerDay)
            let window = try DateRange(from: windowStart, to: now)
            let counts = try await client.healthSync.sync(Set(HealthDataType.allCases), window: window)
            let perTypeSummary = counts.perType
                .map { "\($0.key.rawValue): \($0.value)" }
                .sorted()
                .joined(separator: ", ")
            return "Synced \(counts.total) records.\nPer type: \(perTypeSummary)"

        case .getOauthUrl:
            guard let slug = selectedDeviceProvider?.slug else {
                throw PlaygroundInputError.missingInput("Select a provider first")
            }
            let result = try await client.connection.getOAuthURL(.init(connectionId: slug))
            return prettyJSON(result)

        case .setupMobileSync:
            let result = try await client.device.setupMobileSync(.init(connectionId: Self.playgroundConnectionId))
            return prettyJSON(result)

        case .getDeviceUserStatus:
            let result = try await client.device.getDeviceUserStatus(.init(connectionId: Self.playgroundConnectionId))
            return prettyJSON(result)

        case .getDeviceProviderStatus:
            guard let slug = selectedDeviceProvider?.slug else {
                throw PlaygroundInputError.missingInput("Select a provider first")
            }
            let result = try await client.device.getDeviceProviderStatus(.init(connectionId: slug))
            return prettyJSON(result)

        case .getDeviceProviders:
            let result = try await client.device.getDeviceProviders()
            return prettyJSON(result)

        case .deleteConnection:
            // Exercises the raw `connection.deleteConnection` ATS call only -
            // not the full `client.disconnect()` composition (session
            // teardown + connection delete), which has its own card above.
            let connectionId = connectionIdInput.trimmed
            guard !connectionId.isEmpty else {
                throw PlaygroundInputError.missingInput("Enter a connection id first")
            }
            let result = try await client.connection.deleteConnection(.init(connectionId: connectionId))
            return prettyJSON(result)

        case .getDeviceMetrics:
            let groupCode = selectedDeviceMetricsCode.map { BWell.SearchToken(value: .init(code: $0)) }
            let result = try await client.health.getDeviceMetrics(.init(page: 0, groupCode: groupCode))
            return prettyJSON(result)

        case .getHealthScore:
            let result = try await client.health.getHealthScore()
            return prettyJSON(result)

        case .getBodySystemScore:
            guard let bodySystemId = selectedBodySystemId else {
                throw PlaygroundInputError.missingInput("Select a body system first")
            }
            let result = try await client.health.getBodySystemScore(.init(bodySystemId: bodySystemId))
            return prettyJSON(result)

        case .getDeviceMetricsGroups:
            let result = try await client.health.getDeviceMetricsGroups(.init(page: 0))
            return prettyJSON(result)
        }
    }

    // Playground-only, for visibility while testing - pretty-printed JSON of
    // whatever a card's response holds, not an SDK-wide serialization format.
    private static let playgroundJSONEncoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return encoder
    }()

    private func prettyJSON(_ value: some Encodable) -> String {
        guard let data = try? Self.playgroundJSONEncoder.encode(value),
              let json = String(data: data, encoding: .utf8) else {
            return String(describing: value)
        }
        // Some URL-shaped fields (e.g. Coding.system) come through already
        // containing literal backslash-escaped slashes - purely cosmetic for
        // this display-only view, so unescape for readability.
        return json.replacingOccurrences(of: "\\/", with: "/")
    }
}

private extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}

File: bwell-kotlin-android/app/src/main/java/com/bwell/sampleapp/activities/ui/healthsync/HealthSyncPlaygroundViewModel.kt
package com.bwell.sampleapp.activities.ui.healthsync

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bwell.common.models.domain.data.DeviceProvider
import com.bwell.common.models.domain.data.DeviceProviderList
import com.bwell.common.models.domain.healthdata.healthsummary.devicemetrics.DeviceMetricsGroup
import com.bwell.common.models.domain.healthdata.healthsummary.healthscore.HealthScoreResult
import com.bwell.common.models.responses.BWellResult
import com.bwell.sampleapp.repository.HealthSyncRepository
import com.google.gson.GsonBuilder
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Per-card result - each card runs and renders independently of the others. */
sealed interface PlaygroundCardState {
    data object Idle : PlaygroundCardState
    data object Loading : PlaygroundCardState
    data class Success(val raw: String) : PlaygroundCardState
    data class Error(val message: String) : PlaygroundCardState
}

/** A dropdown option carrying a raw code + its display label - ported from Swift's PlaygroundCodeOption. */
data class PlaygroundCodeOption(val code: String, val display: String?) {
    val label: String get() = if (display != null) "$code - $display" else code
}

data class BodySystemOption(val bodySystemId: String, val title: String?) {
    val label: String get() = title ?: bodySystemId
}

/**
 * Ported from Swift's HealthSyncPlaygroundViewModel. Drives every card in
 * [HealthSyncPlaygroundScreen]: dispatches a run, tracks per-card state, and
 * lazily loads the 3 dynamic dropdowns (device providers, device-metrics
 * codes, body systems) the same way Swift's `.task {}` modifiers do.
 */
class HealthSyncPlaygroundViewModel(private val repository: HealthSyncRepository) : ViewModel() {

    companion object {
        // Matches the internal healthsync-sample's convention for the
        // on-device source's connection id. Hardcoded here, same as Swift's
        // Playground hardcodes "apple_health" for setupMobileSync/
        // getDeviceUserStatus - it is not a user-editable field.
        const val PLAYGROUND_CONNECTION_ID = "health_connect"

        private const val TAG = "HealthSyncPlayground"
    }

    private val _results = MutableStateFlow<Map<HealthSyncPlaygroundEndpoint, PlaygroundCardState>>(emptyMap())
    val results: StateFlow<Map<HealthSyncPlaygroundEndpoint, PlaygroundCardState>> = _results.asStateFlow()

    private val _configured = MutableStateFlow(repository.isHealthSyncConfigured())
    val configured: StateFlow<Boolean> = _configured.asStateFlow()

    private val _deviceProviderOptions = MutableStateFlow<List<DeviceProvider>>(emptyList())
    val deviceProviderOptions: StateFlow<List<DeviceProvider>> = _deviceProviderOptions.asStateFlow()

    private val _selectedDeviceProvider = MutableStateFlow<DeviceProvider?>(null)
    val selectedDeviceProvider: StateFlow<DeviceProvider?> = _selectedDeviceProvider.asStateFlow()

    // Only used by deleteConnection - auto-prefilled from the selected
    // provider's slug, editable, distinct from PLAYGROUND_CONNECTION_ID.
    private val _deleteConnectionIdInput = MutableStateFlow("")
    val deleteConnectionIdInput: StateFlow<String> = _deleteConnectionIdInput.asStateFlow()

    private val _deviceMetricsCodeOptions = MutableStateFlow<List<PlaygroundCodeOption>>(emptyList())
    val deviceMetricsCodeOptions: StateFlow<List<PlaygroundCodeOption>> = _deviceMetricsCodeOptions.asStateFlow()

    private val _selectedDeviceMetricsCode = MutableStateFlow<String?>(null)
    val selectedDeviceMetricsCode: StateFlow<String?> = _selectedDeviceMetricsCode.asStateFlow()

    private val _bodySystemOptions = MutableStateFlow<List<BodySystemOption>>(emptyList())
    val bodySystemOptions: StateFlow<List<BodySystemOption>> = _bodySystemOptions.asStateFlow()

    private val _selectedBodySystemId = MutableStateFlow<String?>(null)
    val selectedBodySystemId: StateFlow<String?> = _selectedBodySystemId.asStateFlow()

    private var attached = false
    private val playgroundJson = GsonBuilder().setPrettyPrinting().create()

    /**
     * One-time setup, idempotent - safe to call from every screen that hosts
     * the Playground (direct or Dashboard tab). Callers must also call
     * [refreshDeviceDataOptions] right after - it is not included here (see
     * its doc).
     */
    fun attach() {
        if (attached) return
        attached = true
        _configured.value = repository.isHealthSyncConfigured()
        loadDeviceProviderOptions()
    }

    /**
     * Not gated like [attach] - call every time the Playground screen
     * (re-)enters composition, e.g. on each Dashboard tab switch back to
     * Playground. GET_DEVICE_METRICS/GET_BODY_SYSTEM_SCORE's own copy
     * promises their dropdowns populate "each time this card appears," and
     * a sync triggered from the Dashboard's "Sync with b.well" button runs
     * through [HealthSyncDashboardViewModel] entirely - this ViewModel has
     * no other way to learn that data now exists.
     */
    fun refreshDeviceDataOptions() {
        loadDeviceMetricsCodeOptions()
        loadBodySystemOptions()
    }

    /** For GET_OAUTH_URL/GET_DEVICE_PROVIDER_STATUS - selection only, never touches deleteConnectionIdInput. */
    fun onSelectDeviceProviderForLookup(provider: DeviceProvider) {
        _selectedDeviceProvider.value = provider
    }

    /**
     * For DELETE_CONNECTION only - prefills the id field from the selected
     * provider's slug. Kept separate from [onSelectDeviceProviderForLookup]:
     * all 3 provider dropdowns previously shared one callback that always
     * overwrote deleteConnectionIdInput, silently discarding a manually
     * edited value whenever a provider was (re-)selected on either of the
     * two unrelated lookup cards.
     */
    fun onSelectDeviceProviderForDelete(provider: DeviceProvider) {
        _selectedDeviceProvider.value = provider
        _deleteConnectionIdInput.value = provider.slug
    }

    fun onDeleteConnectionIdInputChanged(value: String) {
        _deleteConnectionIdInput.value = value
    }

    fun onSelectDeviceMetricsCode(code: String?) {
        _selectedDeviceMetricsCode.value = code
    }

    fun onSelectBodySystem(bodySystemId: String) {
        _selectedBodySystemId.value = bodySystemId
    }

    private fun loadDeviceProviderOptions() {
        viewModelScope.launch {
            // Reachable before login - see guardedCall's doc. The provider
            // picker just stays empty and its dependent cards stay disabled
            // (see isRunDisabled) on failure.
            val result = guardedCall<BWellResult<DeviceProviderList>?>(TAG, "loadDeviceProviderOptions", null) {
                repository.getDeviceProviders()
            } ?: return@launch
            if (result is BWellResult.SingleResource && result.success()) {
                val providers = result.data?.providers.orEmpty()
                _deviceProviderOptions.value = providers
                if (_selectedDeviceProvider.value == null) {
                    providers.firstOrNull()?.let(::onSelectDeviceProviderForDelete)
                }
            }
        }
    }

    /** Dedupes getDeviceMetricsGroups() results by coding.code, sorted - mirrors Swift's loadDeviceMetricsCodeOptions. */
    private fun loadDeviceMetricsCodeOptions() {
        viewModelScope.launch {
            val result = guardedCall<BWellResult<DeviceMetricsGroup>?>(TAG, "loadDeviceMetricsCodeOptions", null) {
                repository.getDeviceMetricsGroups()
            } ?: return@launch
            if (result is BWellResult.ResourceCollection && result.success()) {
                val options = result.data.orEmpty()
                    .mapNotNull { it.coding?.code?.let { code -> PlaygroundCodeOption(code, it.coding?.display) } }
                    .distinctBy { it.code }
                    .sortedBy { it.code }
                _deviceMetricsCodeOptions.value = options
                if (_selectedDeviceMetricsCode.value != null && options.none { it.code == _selectedDeviceMetricsCode.value }) {
                    _selectedDeviceMetricsCode.value = null
                }
            }
        }
    }

    /** Sourced from getHealthScore()'s contributing body systems - mirrors Swift's loadBodySystemOptions. */
    private fun loadBodySystemOptions() {
        viewModelScope.launch {
            val result = guardedCall<BWellResult<HealthScoreResult>?>(TAG, "loadBodySystemOptions", null) {
                repository.getHealthScore()
            } ?: return@launch
            if (result is BWellResult.SingleResource && result.success()) {
                val options = result.data?.resource?.bodySystems.orEmpty()
                    .mapNotNull { it.bodySystemId?.let { id -> BodySystemOption(id, it.title) } }
                _bodySystemOptions.value = options
                if (_selectedBodySystemId.value == null) {
                    options.firstOrNull()?.let { _selectedBodySystemId.value = it.bodySystemId }
                }
            }
        }
    }

    fun run(endpoint: HealthSyncPlaygroundEndpoint) {
        val blockedReason = HealthSyncGating.blockedReason(endpoint, _configured.value)
        if (blockedReason != null) {
            _results.value = _results.value + (endpoint to PlaygroundCardState.Error(blockedReason))
            return
        }
        _results.value = _results.value + (endpoint to PlaygroundCardState.Loading)
        viewModelScope.launch {
            val cardState = try {
                runEndpoint(endpoint)
            } catch (e: Exception) {
                PlaygroundCardState.Error(e.message ?: "Unexpected exception")
            }
            _results.value = _results.value + (endpoint to cardState)

            if (endpoint == HealthSyncPlaygroundEndpoint.SYNC && cardState is PlaygroundCardState.Success) {
                refreshDeviceDataOptions()
            }
        }
    }

    private suspend fun runEndpoint(endpoint: HealthSyncPlaygroundEndpoint): PlaygroundCardState = when (endpoint) {
        HealthSyncPlaygroundEndpoint.CONNECT ->
            repository.connect().toCardState { "connect() completed successfully." }
        HealthSyncPlaygroundEndpoint.DISCONNECT ->
            repository.disconnect().toCardState { "disconnect() completed successfully." }
        HealthSyncPlaygroundEndpoint.REQUEST_PERMISSIONS ->
            repository.requestPermissions().toCardState { "requestPermissions() completed successfully." }
        HealthSyncPlaygroundEndpoint.SYNC ->
            repository.sync().toCardState()
        HealthSyncPlaygroundEndpoint.GET_CURRENT_USER -> {
            val result = repository.currentUser()
            val user = (result as? BWellResult.SingleResource)?.data
            when {
                !result.success() -> PlaygroundCardState.Error(result.error?.message() ?: "Unknown error")
                user == null -> PlaygroundCardState.Success("No active session.")
                else -> PlaygroundCardState.Success(
                    "userId: ${user.userId}\norganizationId: ${user.organizationId}",
                )
            }
        }
        HealthSyncPlaygroundEndpoint.SETUP_MOBILE_SYNC ->
            repository.setupMobileSync(PLAYGROUND_CONNECTION_ID).toCardState()
        HealthSyncPlaygroundEndpoint.GET_DEVICE_USER_STATUS ->
            repository.getDeviceUserStatus(PLAYGROUND_CONNECTION_ID).toCardState()
        HealthSyncPlaygroundEndpoint.GET_OAUTH_URL ->
            requireSelectedProvider()?.let { repository.getOauthUrl(it.slug).toCardState() }
                ?: PlaygroundCardState.Error("Select a provider first")
        HealthSyncPlaygroundEndpoint.GET_DEVICE_PROVIDER_STATUS ->
            requireSelectedProvider()?.let { repository.getDeviceProviderStatus(it.slug).toCardState() }
                ?: PlaygroundCardState.Error("Select a provider first")
        HealthSyncPlaygroundEndpoint.GET_DEVICE_PROVIDERS ->
            repository.getDeviceProviders().toCardState()
        HealthSyncPlaygroundEndpoint.DELETE_CONNECTION -> {
            val connectionId = _deleteConnectionIdInput.value.trim()
            if (connectionId.isEmpty()) {
                PlaygroundCardState.Error("Enter a connection id first")
            } else {
                val outcome = repository.deleteConnection(connectionId)
                if (outcome.success()) {
                    PlaygroundCardState.Success("deleteConnection() completed successfully.")
                } else {
                    PlaygroundCardState.Error(outcome.message())
                }
            }
        }
        HealthSyncPlaygroundEndpoint.GET_DEVICE_METRICS ->
            repository.getDeviceMetrics(_selectedDeviceMetricsCode.value).toCardState()
        HealthSyncPlaygroundEndpoint.GET_DEVICE_METRICS_GROUPS ->
            repository.getDeviceMetricsGroups().toCardState()
        HealthSyncPlaygroundEndpoint.GET_HEALTH_SCORE ->
            repository.getHealthScore().toCardState()
        HealthSyncPlaygroundEndpoint.GET_BODY_SYSTEM_SCORE ->
            _selectedBodySystemId.value?.let { repository.getBodySystemScore(it).toCardState() }
                ?: PlaygroundCardState.Error("Select a body system first")
    }

    private fun requireSelectedProvider(): DeviceProvider? = _selectedDeviceProvider.value

    private fun BWellResult<*>.toCardState(onSuccess: (() -> String)? = null): PlaygroundCardState = when {
        !success() -> PlaygroundCardState.Error(error?.message() ?: "Unknown error")
        onSuccess != null -> PlaygroundCardState.Success(onSuccess())
        this is BWellResult.SingleResource<*> -> PlaygroundCardState.Success(playgroundJson.toJson(data))
        this is BWellResult.ResourceCollection<*> ->
            PlaygroundCardState.Success(playgroundJson.toJson(mapOf("pagingInfo" to pagingInfo, "data" to data)))
        this is BWellResult.SearchResults<*> ->
            PlaygroundCardState.Success(playgroundJson.toJson(mapOf("pagingInfo" to pagingInfo, "data" to data)))
        else -> PlaygroundCardState.Success(playgroundJson.toJson(this))
    }
}
